interop/tapir/src/test/scala/caliban/interop/tapir/ws/ProtocolSpec.scala

```diff
@@ -0,0 +1,100 @@
+package caliban.interop.tapir.ws
+
+import caliban.Value.StringValue
+import caliban.execution.QueryExecution
+import caliban.interop.tapir.TestData.sampleCharacters
+import caliban.interop.tapir.{ TestApi, TestService, WebSocketHooks }
+import caliban.uploads.Uploads
+import caliban.{ GraphQLWSInput, InputValue }
+import zio.clock.Clock
+import zio.console.Console
+import zio.duration.durationInt
+import zio.stream.ZStream
+import zio.test._
+import zio.{ Queue, ZIO }
+
+object ProtocolSpec extends DefaultRunnableSpec {
+
+  def protocol(protocol: Protocol) = for {
+    interpreter <- TestApi.api.interpreter
+    pipe        <- protocol.make(
+                     interpreter,
+                     false,
+                     true,
+                     None,
+                     QueryExecution.Parallel,
+                     WebSocketHooks.empty
+                   )
+  } yield pipe
+
+  override def spec: ZSpec[_root_.zio.test.environment.TestEnvironment, Any] = suite("Websocket Protocol")(
```

Comment:
How about integrating this into the test suite in `TapirAdapterSpec`? It has the advantage of being ran for each http adapter.

Reply:
I was actually hoping to make this test be the de facto test that runs agnostic of the adapter.

Reply:
It's go for testing the internal logic, but each http lib has its quirks so I think it's also useful to run it together. I found several bugs like that (in Caliban and in Tapir).